File: web/src/api/dbCleanup.ts
import request from '../utils/request';

export interface DBCleanupResult {
  orphan_prescriptions: number;
  orphan_items: number;
  orphan_billings: number;
  orphan_user_roles: number;
  orphan_role_permissions: number;
  soft_deleted: Record<string, number>;
  cleaned?: Record<string, number>;
}

/** Scan or cleanup orphan data in the database. */
export function cleanupOrphanData(dryRun: boolean = true) {
  return request.post<DBCleanupResult>(`/db/cleanup?dry_run=${dryRun}`);
}

File: server/handler/db_cleanup.go
package handler

import (
	"log"

	"github.com/callmefisher/menzhen/server/service"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// DBCleanupHandler handles database orphan data cleanup endpoints.
type DBCleanupHandler struct {
	db *gorm.DB
}

// NewDBCleanupHandler creates a new DBCleanupHandler.
func NewDBCleanupHandler(db *gorm.DB) *DBCleanupHandler {
	return &DBCleanupHandler{db: db}
}

// CleanupOrphanData handles POST /api/v1/db/cleanup.
// Query param dry_run=false triggers actual deletion; default is dry_run=true (scan only).
func (h *DBCleanupHandler) CleanupOrphanData(c *gin.Context) {
	dryRun := c.DefaultQuery("dry_run", "true") != "false"

	svc := service.NewDBCleanupService(h.db)

	if dryRun {
		result, err := svc.ScanOrphanData()
		if err != nil {
			log.Printf("[db_cleanup] scan error: %v", err)
			Error(c, 500, "扫描孤立数据失败")
			return
		}
		Success(c, result)
		return
	}

	result, err := svc.CleanupOrphanData()
	if err != nil {
		log.Printf("[db_cleanup] cleanup error: %v", err)
		Error(c, 500, "清理孤立数据失败")
		return
	}
	Success(c, result)
}
